feat: Adding a new controller for Metametrics Data Deletion

NiranjanaBinoy opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
We are adding a new controller, `MetaMetricsDataDeletionController,` and a service class for MetaMetrics data deletion. The service class will have the API interaction for the data deletion, which will create the data deletion regulation and check the status of an existing relation. The controller handles the state update and service class interactions.
The PR also initiates the controller by updating the metamask-controller and updating actions.ts with the new functions.
<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24503?quickstart=1)

## **Related issues**

Fixes:
fixes #24405 
## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

NiranjanaBinoy committed:
adding a new controller and service class for metametrics deletion

NiranjanaBinoy committed:
adding env variable

NiranjanaBinoy committed:
updating .metamaskrc.dist with new variables

NiranjanaBinoy committed:
updating build.yml entries

NiranjanaBinoy committed:
updating build.yml entries to fix circle ci failure

NiranjanaBinoy committed:
updating metamask-controller and action with the new controller

NiranjanaBinoy committed:
updating setupsentry

NiranjanaBinoy committed:
fixing circle ci failure

danjm commented on ui/store/actions.ts:
```diff
@@ -5085,6 +5086,42 @@ export function setName(
   }) as any;
 }
 
+/**
```
Can you add a unit test for each of these?

NiranjanaBinoy replied:
added

danjm commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,138 @@
+import { ObservableStore } from '@metamask/obs-store';
```
We should add some unit tests for this file.

danjm commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  async createMetaMetricsDataDeletionTask(): Promise<DataDeletionResponse> {
+    const segmentSourceId = process.env.SEGMENT_DELETE_API_SOURCE_ID;
```
The definition of these constant variables can happen outside of this method and outside of the class. We don't need to re-define them every time the method is called.

danjm commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,138 @@
+      );
+
+      this.store.updateState({
+        metaMetricsDataDeletionId: data?.data?.regulateId as string,
```
If ` data?.data?.regulateId` is `undefined`, should we return an error?

NiranjanaBinoy replied:
I believe the api will return a regulate ID if the request did not error. So I have made the type to just string than `string | undefined`, and initiated the state variable to empty string.

danjm commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,138 @@
+      const { data } = await fetchDeletionRegulationStatus(
+        this.store.getState().metaMetricsDataDeletionId as string,
+      );
+      const regulation = data?.data?.regulation as Record<string, string>;
```
If `data?.data?.regulation` is `undefined`, should we return an error?

NiranjanaBinoy replied:
I believe the api will return a regulation if we have a valid Id, otherwise, it would probably error. I have added an empty check for `deleteRegulateId` to make sure we have a value before we call the api..

danjm commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,138 @@
+    const segmentSourceId = process.env.SEGMENT_DELETE_API_SOURCE_ID;
+    const segmentRegulationEndpoint = process.env.SEGMENT_REGULATIONS_ENDPOINT;
+
+    if (!segmentSourceId || !segmentRegulationEndpoint) {
```
Maybe this validation is unneeded here, and can just happen within the services?

NiranjanaBinoy replied:
moved

danjm commented on app/scripts/controllers/metametrics-data-deletion/services/services.ts:
```diff
@@ -0,0 +1,48 @@
+    );
+    return response as DeleteRegulationAPIResponse;
+  } catch (error: unknown) {
+    throw new Error('Analytics Deletion Task Error');
```
Any reason not to throw the `error` object received in the catch?

NiranjanaBinoy replied:
Removed the try-catch in the service class, and it is being handled in the controller.

danjm commented on app/scripts/controllers/metametrics-data-deletion/services/services.ts:
```diff
@@ -0,0 +1,48 @@
+  const segmentRegulationEndpoint = process.env.SEGMENT_REGULATIONS_ENDPOINT;
+
+  try {
+    const response: unknown = await fetchWithTimeout(
```
I wonder if we can type the response in a way other that `unknown` :thinking:

NiranjanaBinoy replied:
The type was explicitly mentioned as it is erroring while converting the `Response` type returned from `fetchWithTimeout ` to DeleteRegulationAPIResponse. We can also typecast while returning the response like this `response as unknown as DeleteRegulationAPIResponse`

danjm reviewed (changes requested):
First review complete. I have left some comments

NiranjanaBinoy committed:
adding new state variable and addressing review comments

NiranjanaBinoy committed:
updating the empty check

NiranjanaBinoy committed:
checking for the response being empty

NiranjanaBinoy committed:
converting the response to json

NiranjanaBinoy committed:
removing the return

Gudahtt commented on .metamaskrc.dist:
```diff
@@ -6,6 +6,8 @@ INFURA_PROJECT_ID=00000000000
 
 ;PASSWORD=METAMASK PASSWORD
 ;SEGMENT_WRITE_KEY=
+;SEGMENT_DELETE_API_SOURCE_ID=
```
Could you clarify what these environment variables are?

It was my understanding that we're using our own API for this, not using Segment directly. Our API will also be requesting deletion from more than just Segment. If that's the case, making these environment variables named `SEGMENT_` seems inaccurate

NiranjanaBinoy replied:
updated

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+    initState: MetaMetricsDataDeletionState;
+    metaMetricsController: MetaMetricsController;
+  }) {
+    this.store = new ObservableStore({
```
It looks like you're building an older extension style controller. These are deprecated, we should be using BaseControllerV2 instead to build controllers.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+    metaMetricsController,
+  }: {
+    initState: MetaMetricsDataDeletionState;
+    metaMetricsController: MetaMetricsController;
```
Passing controller instances is also a deprecated pattern. Instead we have been using the controller messenger, or for legacy controllers, passing controller methods in as constructor parameters.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+  data: RegulationId | CurrentRegulationStatus;
+};
+
+export enum DeleteRegulationStatus {
```
Are there API docs somewhere that can explain what status exist, and what they mean?

NiranjanaBinoy replied:
The existing statuses are mentioned here:

https://docs.segmentapis.com/tag/Deletion-and-Suppression#operation/listRegulationsFromSource

Gudahtt replied:
Could you update the comment to reference these docs?

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+  fetchDeletionRegulationStatus,
+} from './services/services';
+
+export type DataDeleteDate = string;
```
Nit: Ideally each type would have a TSDoc block that explains what the type is for. This is especially useful because the descriptions show up in Intellisense tooltips in supporting editors such as VSCode

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,138 @@
+    this.metaMetricsController = metaMetricsController;
+  }
+
+  _formatDeletionDate() {
```
This is for the internal controller state right? It would seem more practical to store a timestamp instead of a formatted string. Controller state isn't meant for displaying directly to users.

Dates need to be localized, so we'd end up having to parse it into a timestamp/date and re-format it for display either way.

NiranjanaBinoy replied:
Update the date to be a timestamp.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+    return `${day}/${month}/${year}`;
+  }
+
+  async createMetaMetricsDataDeletionTask(): Promise<void> {
```
Nit: TSDoc blocks here would be great, to explain what they are for

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+    } catch (error) {
+      const errorMessage =
+        error instanceof Error ? error.message : JSON.stringify(error ?? '');
+      throw new Error(`Metametrics Data Deletion Error: ${errorMessage}`);
```
What is the purpose of this `try..catch` block? Perhaps we could let this bubble up directly instead? Likely the stack trace would make clear where it came from.

If we do want to explicitly wrap the error, I'm sure we can find a better way of doing that. Handling "thrown strings" is pretty unconventional, as is `JSON.stringify`-ing error objects.

Gudahtt replied:
The same question applies to the `try..catch` block in `checkDataDeletionTaskStatus`

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+
+export type RegulationId = Record<string, string>;
+export type CurrentRegulationStatus = Record<string, Record<string, string>>;
+export type DeleteRegulationAPIResponse = {
```
We're creating an import cycle here because this type is used by `services.ts`, but we're importing `services` here. We could eliminate this cycle by moving these types to a separate file, e.g. `types.ts`

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  async createMetaMetricsDataDeletionTask(): Promise<void> {
+    const { metaMetricsId } = this.metaMetricsController.store.getState();
```
We can state the type we expect here, so that we don't need to cast it later on line 91.

```suggestion
    const { metaMetricsId }: { metaMetricsId: string | null } = this.metaMetricsController.store.getState();
```

See these docs for why we'd recommend a type annotation over a cast: https://github.com/MetaMask/contributor-docs/blob/main/docs/typescript.md#type-annotations

This specific case (interfacing with an untyped module) isn't covered, but the advice there is relevant here:

>Compared to type assertions, type annotations are more responsive to code drift. If the assignee's type becomes incompatible with the assigned type annotation, the compiler will raise a type error, whereas in most cases a type assertion will still suppress the error.

Gudahtt replied:
Oh actually, wait, extension is on TypeScript v5. We can use `satisfies`, that's even better

```suggestion
    const { metaMetricsId } = this.metaMetricsController.store.getState() satisfies { metaMetricsId: string | null };
```

NiranjanaBinoy replied:
`metaMetricsId` is added as a property for this class, and then a null check is added before calling the status check function. this eliminated the use of type casting.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+        metaMetricsId as string,
+      );
+      this.store.updateState({
+        metaMetricsDataDeletionId: data?.regulateId as string,
```
I see that `checkDataDeletionTaskStatus` has a couple of similar cases, where we're making assumptions about the data that aren't true of the type we're using.

NiranjanaBinoy replied:
The type casting had to be done because of the response type defined previously, which was used by both methods; I updated that response type and removed the typecasting.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+import { ObservableStore } from '@metamask/obs-store';
+import MetaMetricsController from '../metametrics';
+import {
```
Nit: The way we have been building external API service modules recently is to construct them as classes, then pass them into the controller as constructor parameters (either that or use the messenger to communicate with them, which is planned but hasn't been done yet, see https://github.com/MetaMask/decisions/blob/main/decisions/core/0001-messaging-non-controllers.md)

Injecting the services rather than importing them would make it easier to mock out the service in unit tests

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+import {
+  createDataDeletionRegulationTask,
+  fetchDeletionRegulationStatus,
+} from './services/services';
```
Nit: The way we have been building external API service modules recently is to construct them as classes, then pass them into the controller as constructor parameters (either that or use the messenger to communicate with them, which is planned but hasn't been done yet, see https://github.com/MetaMask/decisions/blob/main/decisions/core/0001-messaging-non-controllers.md)

Injecting the services rather than importing them would make it easier to mock out the service in unit tests

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/services/services.ts:
```diff
@@ -0,0 +1,45 @@
+
+const fetchWithTimeout = getFetchWithTimeout();
+
+export async function createDataDeletionRegulationTask(
```
I would recommend converting these two methods into a class instead, so that we can more easily incorporate a retry policy (as an instance variable, because it's stateful).

You can see an example of this here: https://github.com/MetaMask/core/blob/main/packages/assets-controllers/src/token-prices-service/codefi-v2.ts

Copying the same basic retry policy should be fine at least for the status check. It would make this feature more resilient to network connectivity issues and sporadic API downtime.

I'm less sure about the POST request. Not sure how it acts when submitting the same deletion request twice, that might affect how we deal with retries.

Gudahtt commented on ui/store/actions.ts:
```diff
@@ -5085,6 +5086,39 @@ export function setName(
+      await submitRequestToBackground('createMetaMetricsDataDeletionTask');
+      await forceUpdateMetamaskState(dispatch);
+    } catch (error) {
+      logErrorWithMessage(error);
```
Why are we suppressing errors here?

Gudahtt replied:
Likewise for `checkDataDeletionTaskStatus`

Gudahtt commented on ui/store/actions.ts:
```diff
@@ -5085,6 +5086,39 @@ export function setName(
+/**
+ * To check the status of the current delete regulation.
+ */
+export async function checkDataDeletionTaskStatus(): Promise<DataDeletionResponse> {
```
Hmm, the return type seems wrong here. There is no `DataDeletionResponse` returned

NiranjanaBinoy replied:
updated

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,125 @@
+    }
+  }
+
+  async checkDataDeletionTaskStatus(): Promise<void> {
```
Nit: maybe `updateDataDeletionTaskStatus` would be a clearer name here? "check" seems like it might return the status, which this does not.

NiranjanaBinoy replied:
updated

Gudahtt commented on ui/store/actions.ts:
```diff
@@ -5085,6 +5086,39 @@ export function setName(
+    await submitRequestToBackground<DataDeletionResponse>(
+      'checkDataDeletionTaskStatus',
+    );
+    await forceUpdateMetamaskState(dispatch);
```
This seems to be broken (`dispatch` is unbound)

NiranjanaBinoy replied:
updated

NiranjanaBinoy committed:
updating the controller pattern, renaming the env variable, updating some state files to fix circle ci failure'

NiranjanaBinoy committed:
addressing review comments

NiranjanaBinoy committed:
fixing the actions

NiranjanaBinoy committed:
lint fix

NiranjanaBinoy committed:
updating env variable and e2e errors

Gudahtt committed:
feat: Add retry policy to the data deletion service (#24716)

NiranjanaBinoy committed:
updating teh actions and test

NiranjanaBinoy committed:
updating teh date format

NiranjanaBinoy committed:
fixing rebase conflict

Gudahtt committed:
Add additional data deletion service configuration options (#24806)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

The `circuitBreakDuration`, `degradedThreshold`, and `timeout` are now
configurable for the `DataDeletionService`. This configuration is relied
upon in the unit tests, to be added in a later PR.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24806?quickstart=1)

## **Related issues**

This is an improvement for the PR #24503

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

NiranjanaBinoy committed:
updating the error states for e2e

Gudahtt committed:
refactor: Create dedicated services directory (#24863)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

The `DataDeletionService` has been moved to a `services` directory,
independent of controllers. This was done to reinforce that services are
independent of controllers, and can be used outside of them.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24863?quickstart=1)

This is an improvement for the PR
https://github.com/MetaMask/metamask-extension/pull/24503

N/A

N/A

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

Gudahtt committed:
refactor: Decouple the data deletion controller (#24870)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

The `DataDeletionController` has been decoupled from the
`MetaMetricsController` by updating the constructor parameters to expect
just a function for getting the MetaMetrics ID, rather than the entire
`MetaMetricsController` state. This vastly simplifies the API surface,
making it easier to audit and test. It also prevents the
`DataDeletionController` from having the capability to make changes to
the `MetaMetricsController` state.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24870?quickstart=1)

## **Related issues**

This is an improvement for #24503

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

Gudahtt committed:
test: Add a MetaMetricsDataDeletionController test (#24879)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

A single test has been added, along with a helper setup function that is
designed to be used for further unit tests as well.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24879?quickstart=1)

This is an improvement for the PR
https://github.com/MetaMask/metamask-extension/pull/24503

N/A

N/A

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

Gudahtt committed:
Add tests for `createDataDeletionRegulationTask` (#24816)

Add unit tests for the `createDataDeletionRegulationTask` method of
`DataDeletionService`.

One of the two service methods is now fully tested. The remaining method
will be tested in a later PR.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24816?quickstart=1)

This is an improvement for the PR #24503

N/A

N/A

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Niranjana Binoy <43930900+NiranjanaBinoy@users.noreply.github.com>

NiranjanaBinoy committed:
adding a new property to state

NiranjanaBinoy committed:
updating test files of the controller and service class

NiranjanaBinoy committed:
fixing lint error

NiranjanaBinoy commented on builds.yml:
```diff
@@ -23,8 +23,12 @@ buildTypes:
+      - SEGMENT_REGULATIONS_PROD_ENDPOINT
       - INFURA_ENV_KEY_REF: INFURA_PROD_PROJECT_ID
       - SEGMENT_WRITE_KEY_REF: SEGMENT_PROD_WRITE_KEY
+      - SEGMENT_DELETE_API_SOURCE_ID_REF: SEGMENT_DELETE_API_PROD_SOURCE_ID
```
changed

NiranjanaBinoy committed:
fixing rebase errors

NiranjanaBinoy committed:
rebase fix

NiranjanaBinoy committed:
fixing lint error

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,220 @@
+ * @type DeleteRegulationStatus
+ */
+export enum DeleteRegulationStatus {
+  FAILED = 'FAILED',
+  FINISHED = 'FINISHED',
+  INITIALIZED = 'INITIALIZED',
+  INVALID = 'INVALID',
+  NOT_SUPPORTED = 'NOT_SUPPORTED',
+  PARTIAL_SUCCESS = 'PARTIAL_SUCCESS',
+  RUNNING = 'RUNNING',
+  UNKNOWN = 'UNKNOWN',
```
nit: I believe we generally use PascalCase for enum keys.
(e.g. `DeleteRegulationStatus.PartialSuccess` not `DeleteRegulationStatus.PARTIAL_SUCCESS`)

```suggestion
  Failed = 'FAILED',
  Finished = 'FINISHED',
  Initialized = 'INITIALIZED',
  Invalid = 'INVALID',
  NotSupported = 'NOT_SUPPORTED',
  PartialSuccess = 'PARTIAL_SUCCESS',
  Running = 'RUNNING',
  Unknown = 'UNKNOWN',
```

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,220 @@
+  }: {
+    dataDeletionService: PublicInterface<DataDeletionService>;
+    messenger: MetaMetricsDataDeletionControllerMessenger;
+    state?: MetaMetricsDataDeletionState;
```
```suggestion
    state?: Partial<MetaMetricsDataDeletionState>;
```

It should be safe to supply an incomplete set of state properties, since we're populating the state object with `defaultState` first.

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,220 @@
+    dataDeletionService: PublicInterface<DataDeletionService>;
+    messenger: MetaMetricsDataDeletionControllerMessenger;
+    state?: MetaMetricsDataDeletionState;
+    getMetaMetricsId: () => string | null;
```
Could this be replaced with a messenger action? What's the upstream source of this callback?

MajorLift replied:
Looks like this is coming from `MetaMetricsController` which uses `ObservableStore`. A constructor option callback should work for now.

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,220 @@
+
+  /**
+   * Creating the delete regulation using source regulation
+   *
```
nit:
```suggestion
```

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,220 @@
+      throw new Error('Delete Regulation id not found');
+    }
+
+    const { data } =
+      await this.#dataDeletionService.fetchDeletionRegulationStatus(
+        deleteRegulationId,
+      );
+
+    const regulation = data?.regulation;
```
nit: According to the types, it looks like `data` and `regulation` should never be undefined. Is there's a possibility that they're undefined at runtime? If so, do we need to explicitly handle that case?

```suggestion
    const { data: { regulation } } =
      await this.#dataDeletionService.fetchDeletionRegulationStatus(
        deleteRegulationId,
      );

```

NiranjanaBinoy replied:
As long as the api fetch is successful, it should be returning `regulation`

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,220 @@
+    const regulation = data?.regulation;
+    this.update((state) => {
+      state.metaMetricsDataDeletionStatus =
+        regulation.overallStatus as DeleteRegulationStatus;
```
```suggestion
        regulation.overallStatus;
```

This as assertion should be removed by supplying better typing for `CurrentRegulationStatus` (the return type of `fetchDeletionRegulationStatus`).

If we can always expect `CurrentRegulationStatus` to have the nested property `data.regulation.overallStatus` and its type to be `DeleteRegulationStatus`, that should be reflected in the type definition.

```ts
export type CurrentRegulationStatus = {
  data: { regulation: { overallStatus: DeleteRegulationStatus } };
};
```

MajorLift commented on ui/store/actions.ts:
```diff
@@ -5152,6 +5152,39 @@ export function setName(
+/**
+ * To create a data deletion regulation for MetaMetrics data deletion
+ */
+export function createMetaMetricsDataDeletionTask(): Promise<void> {
+  return async () => {
+    await submitRequestToBackground<void>('createMetaMetricsDataDeletionTask');
+  };
+}
```
If this function returns a promise, shouldn't it be async?

```suggestion
export async function createMetaMetricsDataDeletionTask(): Promise<void> {
  return async () => {
    await submitRequestToBackground<void>('createMetaMetricsDataDeletionTask');
  };
}
```

I'm not sure I understand the purpose of wrapping the return statement in a callback. Is there a reason we can't do one of the following?

```suggestion
export function createMetaMetricsDataDeletionTask(): void {
  return await submitRequestToBackground<void>('createMetaMetricsDataDeletionTask');
}
```

```suggestion
export async function createMetaMetricsDataDeletionTask(): Promise<void> {
  return submitRequestToBackground<void>('createMetaMetricsDataDeletionTask');
}
```

The same applies to `updateDataDeletionTaskStatus` and `setHasMetaMetricsDataRecorded`.

NiranjanaBinoy committed:
addressing review comments

NiranjanaBinoy committed:
updating the actions file

Gudahtt commented on app/scripts/services/data-deletion-service.ts:
```diff
@@ -0,0 +1,274 @@
+        },
+      ),
+    );
+    return response.json();
```
It looks like the only property we use from this response is `data.regulation.overallStatus. Perhaps we should return that directly here, rather than passing along the entire response object that we have no use for.

That would also mean validating that the `overallStatus` property exists on the response. Right now we're not doing any validation, so this would fail if the response was shaped differently or was missing that value.

Gudahtt commented on app/scripts/services/data-deletion-service.ts:
```diff
@@ -0,0 +1,274 @@
+        },
+      ),
+    );
+    return response.json();
```
The only data we use from this method is `data.regulateId`. We should consider returning that directly.

We should also validate that the property exists. The controller would break if it was missing (it would end up setting `metaMetricsDataDeletionId` to `undefined`, and causing a "cannot call .length of undefined" error when the task status was checked)

Gudahtt commented on app/scripts/services/data-deletion-service.ts:
```diff
@@ -0,0 +1,274 @@
+        },
+      ),
+    );
+    return response.json();
```
Note that `fetchWithTimeout` doesn't validate that the response was successful. We might be returning an error message response here. We can guard against this with a check like this:

```
  if (!response.ok) {
    throw new Error(
      `Fetch failed with status '${response.status}' for request '${String(
        request,
      )}'`,
    );
  }
```

Gudahtt replied:
Likewise for the status network call as well

Gudahtt commented on app/scripts/services/data-deletion-service.test.ts:
```diff
@@ -0,0 +1,1209 @@
+          circuitBreakDuration,
+        });
+        const fetchOperation = () =>
+          dataDeletionService.createDataDeletionRegulationTask(
```
This is the wrong method being tested here. This is in the `fetchDeletionRegulationStatus` describe block.

Gudahtt commented on app/scripts/services/data-deletion-service.test.ts:
```diff
@@ -0,0 +1,1209 @@
+  });
+
+  describe('fetchDeletionRegulationStatus', () => {
+    it('fetching the delete regulation status', async () => {
```
Nit: Test description grammar
```suggestion
    it('fetches the delete regulation status', async () => {
```

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,222 @@
+   *
+   * @param record - boolean value to determine data is being recorded.
+   */
+  setHasMetaMetricsDataRecorded(record: boolean): void {
```
What is this method for? Generally speaking, "setter" methods on controllers are something we should be avoiding, they are an anti-pattern. Controllers should be responsible for guaranteeing the correctness/consistency of their own state, and setters make that impossible (the caller can use them to set state incorrectly).

Gudahtt replied:
I see that in #24571, this is called when the MetaMetrics toggle is enabled. Even more concerningly, we're relying on the UI to tell us when this happens.

Rather than asking the caller to know exactly when to call this, the deletion controller should listen for state changes of the toggle itself. Then it can react to the toggle being flipped immediately, without any caller needing to know about it.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,222 @@
+   */
+  setHasMetaMetricsDataRecorded(record: boolean): void {
+    this.update((state) => {
+      state.hasMetaMetricsDataRecorded = record;
```
Why do we need to know when MetaMetrics data has been recorded in the first place?

I'm guessing we wanted to track this because we want to re-enable the "Delete data" button at that point. But... could we instead delete the state about the deletion request? Once metrics are enabled again, it doesn't seem like we use the deletion request data anywhere.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,222 @@
+    this.update((state) => {
+      state.metaMetricsDataDeletionId = data?.regulateId;
+      state.metaMetricsDataDeletionDate = Date.now();
+      return state;
```
We don't need to return `state` here. `state` is an Immer draft object, which records and applies any mutations we make to it.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,222 @@
+      await this.#dataDeletionService.createDataDeletionRegulationTask(
+        metaMetricsId,
+      );
+    this.update((state) => {
```
We should clear `metaMetricsDataDeletionStatus` here as well. If it's set, then it must have been pertaining to a previous deletion request, so it would be inaccurate at this point.

The same goes for `hasMetaMetricsDataRecorded` as well, though in my other comment I suggest removing that state altogether.

danjm replied:
@NiranjanaBinoy was this comment addressed? I see that `this.updateDataDeletionTaskStatus()` is called, but it does not clear the status, rather it updates it

NiranjanaBinoy replied:
Yes, now `metaMetricsDataDeletionStatus` will get updated with the current/newest request status with this call to `this.updateDataDeletionTaskStatus()`.

Gudahtt commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,222 @@
+};
+
+const defaultState: MetaMetricsDataDeletionState = {
+  metaMetricsDataDeletionId: '',
```
Nit: Perhaps `null` would be a better default here than an empty string. That would let us lean more on the type system to ensure we're checking for the case where it's not set.

I see below we do have a check for the case where this is empty (`if (deleteRegulationId.length === 0)`), but we had to remember to do that ourselves. If this defaulted to `null`, the type system would ensure we _can't_ forget about it.

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -182,6 +182,10 @@ export const SENTRY_BACKGROUND_STATE = {
     dataCollectionForMarketing: false,
   },
+  MetaMetricsDataDeletionController: {
+    metaMetricsDataDeletionId: false,
+    metaMetricsDataDeletionDate: false,
```
This is inconsistent with the controller metadata state. The mask suggests that these should be hidden from debug state snapshots, but the metadata suggests they should be included (because `anonymous` is set to `true`).

I believe these should be safe to include, as they only identify the MetaMetrics ID, which itself should only be associated with anonymous data (and in general we set a slightly higher bar for anonymizing metrics state than we do Sentry state)

```suggestion
    metaMetricsDataDeletionId: true,
    metaMetricsDataDeletionDate: true,
```

NiranjanaBinoy committed:
addressing review comments

NiranjanaBinoy committed:
fixing unit test and e2e failures

NiranjanaBinoy committed:
fixing e2e failures

NiranjanaBinoy committed:
removed the setHasMetaMetricsDataRecorded instead update the variable while creating delete regulation

NiranjanaBinoy committed:
fixing e2e failure

NiranjanaBinoy committed:
correcting variable name

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.test.ts:
```diff
@@ -0,0 +1,191 @@
+      expect(
+        dataDeletionService.fetchDeletionRegulationStatus,
+      ).toHaveBeenCalledTimes(1);
+      expect(controller.state).toMatchObject({
+        metaMetricsDataDeletionId: mockTaskId,
+        metaMetricsDataDeletionDate: expect.any(Number),
+        metaMetricsDataDeletionStatus: 'UNKNOWN',
+        participateInMetricsDuringDeletion: true,
+      });
```
Are we sure we want to use `toMatchObject` instead of `toStrictEqual` throughout the tests in this file?

`toMatchObject` matches a subset of the object's properties, so the tests would pass even if `controller.state` were e.g. `{ metaMetricsDataDeletionId: mockTaskId }`.

```suggestion
      expect(controller.state).toStrictEqual({
        metaMetricsDataDeletionId: mockTaskId,
        metaMetricsDataDeletionDate: expect.any(Number),
        metaMetricsDataDeletionStatus: 'UNKNOWN',
        participateInMetricsDuringDeletion: true,
      });
```

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.test.ts:
```diff
@@ -0,0 +1,191 @@
+      await expect(
+        controller.createMetaMetricsDataDeletionTask(),
+      ).rejects.toThrow();
+      expect(controller.state).toMatchObject({
+        metaMetricsDataDeletionId: null,
+        metaMetricsDataDeletionDate: expect.any(Number),
+        participateInMetricsDuringDeletion: null,
+      });
```
Are we sure we want to use `toMatchObject` instead of `toStrictEqual` here? This test will pass even if `controller.state` includes the `metaMetricsDataDeletionStatus` property.

I think it would be safer to use `toStrictEqual` throughout this test file as well, to ensure that we don't get passing tests in the presence of superfluous state properties.

```suggestion
      expect(controller.state).toStrictEqual({
        metaMetricsDataDeletionId: null,
        metaMetricsDataDeletionDate: expect.any(Number),
        participateInMetricsDuringDeletion: null,
      });
```

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.test.ts:
```diff
@@ -0,0 +1,191 @@
+/**
+ * Setup a test controller instance.
+ *
+ * @param options - Setup options.
+ * @param options.options - Controller constructor options.
+ * @param options.dataDeletionService - A partial mock of the DataDeletionService.
+ * @returns The test controller, a messenger instance, and related mocks.
+ */
+function setupController({
+  options,
+  dataDeletionService,
+}: {
+  options?: Partial<
+    ConstructorParameters<typeof MetaMetricsDataDeletionController>[0]
+  >;
+  dataDeletionService?: Partial<
+    ConstructorParameters<
+      typeof MetaMetricsDataDeletionController
+    >[0]['dataDeletionService']
+  >;
```
nit: Is there a reason we have the `dataDeletionService` argument outside of the options bag? It might be confusing to have the controller constructor options scattered in different locations.

Also, it would be helpful to provide JSDoc entries for the constructor options.

```suggestion
 * @param options - Setup options.
 * @param options.options - Controller constructor options.
 * @param options.options.dataDeletionService - A partial mock of the DataDeletionService.
 * @param options.options.getMetaMetricsId - A mock function that returns a mocked MetaMetrics ID.
 * @param options.options.getParticipateInMetrics - A mocked implementation of the `getParticipateInMetrics` method.
 * @returns The test controller, a messenger instance, and related mocks.
 */
function setupController({
  options,
}: {
  options?: Partial<
    ConstructorParameters<typeof MetaMetricsDataDeletionController>[0]
  >;
```

NiranjanaBinoy committed:
updating the timestamp and removing teh boolean prop from state

NiranjanaBinoy committed:
addressing review comments

NiranjanaBinoy committed:
removing the background field for error.spec.js

NiranjanaBinoy committed:
fixing error.spec

mikesposito commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,189 @@
+  metaMetricsDataDeletionStatus?: DeleteRegulationStatus;
+};
+
+const defaultState: MetaMetricsDataDeletionState = {
+  metaMetricsDataDeletionId: null,
+  metaMetricsDataDeletionTimestamp: 0,
+};
```
nit: in some controllers we started preferring functions over variables for default states, to avoid changing the value at runtime:

```typescript
const getDefaultMetaMetricsDataDeletionState =
  (): MetaMetricsDataDeletionState => ({
    metaMetricsDataDeletionId: null,
    metaMetricsDataDeletionTimestamp: 0,
  });
```

NiranjanaBinoy committed:
updated package.json after rebase

NiranjanaBinoy committed:
updating lavamoat

hjetpoluru reviewed (dismissed):
Manual testing looking good.

NiranjanaBinoy committed:
updating teh default state

NiranjanaBinoy committed:
fixing jest config file

MajorLift commented on app/scripts/controllers/metametrics-data-deletion/metametrics-data-deletion.ts:
```diff
@@ -0,0 +1,191 @@
+import {
+  BaseController,
+  RestrictedControllerMessenger,
+} from '@metamask/base-controller';
+import type { DataDeletionService } from '../../services/data-deletion-service';
+import { DeleteRegulationStatus } from '../../../../shared/constants/metametrics';
+
+// Unique name for the controller
+const controllerName = 'MetaMetricsDataDeletionController';
+
+/**
+ * Get a type representing the public interface of the given type. The
+ * returned type will have all public properties, but will omit private
+ * properties.
+ *
+ * @template Interface - The interface to return a public representation of.
+ */
+type PublicInterface<Interface> = Pick<Interface, keyof Interface>;
```
Now that `@metamask/utils` is updated to `>9.1.0` in develop, we could deduplicate the `PublicInterface` type here:

```suggestion
import {
  BaseController,
  RestrictedControllerMessenger,
} from '@metamask/base-controller';
import type { PublicInterface } from '@metamask/utils';
import type { DataDeletionService } from '../../services/data-deletion-service';
import { DeleteRegulationStatus } from '../../../../shared/constants/metametrics';

// Unique name for the controller
const controllerName = 'MetaMetricsDataDeletionController';
```

MajorLift commented on shared/constants/metametrics.ts:
```diff
@@ -848,3 +848,18 @@ export enum MetaMetricsEventUiCustomization {
+ * The status on which to filter the returned regulations.
+ * Mentioned here: https://docs.segmentapis.com/tag/Deletion-and-Suppression#operation/listRegulationsFromSource
+ */
+export enum DeleteRegulationStatus {
+  failed = 'FAILED',
+  finished = 'FINISHED',
+  initialized = 'INITIALIZED',
+  invalid = 'INVALID',
+  notSupported = 'NOT_SUPPORTED',
+  partialSuccess = 'PARTIAL_SUCCESS',
+  running = 'RUNNING',
+  unknown = 'UNKNOWN',
+}
```
nit: we usually use PascalCase for enum keys (e.g. `DeleteRegulationStatus.Failed` vs. `DeleteRegulationStatus.failed`).

```suggestion
export enum DeleteRegulationStatus {
  Failed = 'FAILED',
  Finished = 'FINISHED',
  Initialized = 'INITIALIZED',
  Invalid = 'INVALID',
  NotSupported = 'NOT_SUPPORTED',
  PartialSuccess = 'PARTIAL_SUCCESS',
  Running = 'RUNNING',
  Unknown = 'UNKNOWN',
}
```

NiranjanaBinoy committed:
addressing review comments

MajorLift reviewed (dismissed):
LGTM! Great work on this :)

danjm reviewed (changes requested):
left one question, and merge conflicts need to be resolved

hjetpoluru reviewed (approved):
LGTM
